File: LC/[1778] Shortest Path In A Hidden Grid.py
from collections import deque


class Solution:
    def findShortestPath(self, master: 'GridMaster') -> int:
        # direction, row change, column change, reverse direction
        directions = [
            ('U', -1, 0, 'D'),
            ('D', 1, 0, 'U'),
            ('L', 0, -1, 'R'),
            ('R', 0, 1, 'L'),
        ]

        reachable = {(0, 0)}
        target = (0, 0) if master.isTarget() else None

        # Frame: (row, col, next direction index, move back to parent)
        stack = [(0, 0, 0, None)]

        while stack:
            r, c, i, back = stack[-1]

            if i == 4:
                stack.pop()
                if back is not None:
                    master.move(back)
                continue

            # Resume with the next direction after exploring this one.
            stack[-1] = (r, c, i + 1, back)
            direction, dr, dc, reverse = directions[i]
            nr, nc = r + dr, c + dc

            if (nr, nc) in reachable:
                continue
            if not master.canMove(direction):
                continue

            master.move(direction)
            reachable.add((nr, nc))

            if master.isTarget():
                target = (nr, nc)

            stack.append((nr, nc, 0, reverse))

        if target is None:
            return -1

        # BFS over the discovered grid.
        queue = deque([(0, 0, 0)])
        reachable.remove((0, 0))  # Also use this set as BFS's unvisited set.

        while queue:
            r, c, distance = queue.popleft()

            if (r, c) == target:
                return distance

            for _, dr, dc, _ in directions:
                neighbor = (r + dr, c + dc)
                if neighbor in reachable:
                    reachable.remove(neighbor)
                    queue.append((*neighbor, distance + 1))

        return -1

File: LC/[1778] Shortest Path In A Hidden Grid.java
import java.util.*;

class Solution {
    private static final char[] DIR = {'U', 'D', 'L', 'R'};
    private static final char[] BACK = {'D', 'U', 'R', 'L'};
    private static final int[][] DELTA = {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    private static class Frame {
        int row, col;
        int nextDir;
        int enteredBy;

        Frame(int row, int col, int enteredBy) {
            this.row = row;
            this.col = col;
            this.enteredBy = enteredBy;
        }
    }

    // Encode two signed coordinates into one unique long.
    private long key(int row, int col) {
        return ((long) row << 32) | (col & 0xffffffffL);
    }

    public int findShortestPath(GridMaster master) {
        Set<Long> reachable = new HashSet<>();
        reachable.add(key(0, 0));

        Long target = master.isTarget() ? key(0, 0) : null;

        Deque<Frame> stack = new ArrayDeque<>();
        stack.push(new Frame(0, 0, -1));

        // 1. Explore and map the grid with DFS.
        while (!stack.isEmpty()) {
            Frame current = stack.peek();

            if (current.nextDir == 4) {
                stack.pop();
                if (current.enteredBy != -1) {
                    master.move(BACK[current.enteredBy]);
                }
                continue;
            }

            int d = current.nextDir++;
            int nr = current.row + DELTA[d][0];
            int nc = current.col + DELTA[d][1];
            long nextKey = key(nr, nc);

            if (reachable.contains(nextKey) || !master.canMove(DIR[d])) {
                continue;
            }

            master.move(DIR[d]);
            reachable.add(nextKey);

            if (master.isTarget()) {
                target = nextKey;
            }

            stack.push(new Frame(nr, nc, d));
        }

        if (target == null) {
            return -1;
        }

        // 2. BFS on the discovered grid.
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[]{0, 0, 0}); // row, col, distance
        reachable.remove(key(0, 0));

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int row = current[0];
            int col = current[1];
            int distance = current[2];

            if (key(row, col) == target.longValue()) {
                return distance;
            }

            for (int[] delta : DELTA) {
                int nr = row + delta[0];
                int nc = col + delta[1];

                // Removing on enqueue also marks the cell as visited.
                if (reachable.remove(key(nr, nc))) {
                    queue.offer(new int[]{nr, nc, distance + 1});
                }
            }
        }

        return -1;
    }
}